server/src/main/java/org/apache/druid/segment/realtime/appenderator/UnifiedIndexerAppenderatorsManager.java

```diff
@@ -248,6 +289,50 @@ private AppenderatorConfig rewriteAppenderatorConfigMemoryLimits(AppenderatorCon
     return new MemoryParameterOverridingAppenderatorConfig(baseConfig, perWorkerLimit);
   }
 
+  private class DatasourceBundle
+  {
+    private final SinkQuerySegmentWalker walker;
+    private final Map<String, List<Appenderator>> taskAppenderatorMap;
+
+    public DatasourceBundle(
+        String dataSource
+    )
+    {
+      this.taskAppenderatorMap = new HashMap<>();
+
+      VersionedIntervalTimeline<String, Sink> sinkTimeline = new VersionedIntervalTimeline<>(
+          String.CASE_INSENSITIVE_ORDER
+      );
+      this.walker = new SinkQuerySegmentWalker(
+          dataSource,
+          sinkTimeline,
+          objectMapper,
+          serviceEmitter,
+          queryRunnerFactoryConglomerateProvider.get(),
+          queryExecutorService,
+          Preconditions.checkNotNull(cache, "cache"),
+          cacheConfig,
+          cachePopulatorStats
+      );
+    }
+
+    public SinkQuerySegmentWalker getWalker()
+    {
+      return walker;
+    }
+
+    public void addAppenderator(String taskId, Appenderator appenderator)
+    {
+      taskAppenderatorMap.computeIfAbsent(
+          taskId,
+          (myTaskId) -> {
+            return new ArrayList<>();
+          }
+      );
+      taskAppenderatorMap.get(taskId).add(appenderator);
```

Comment:
I think the below is better to avoid an extra lookup on the map.

```
taskAppenderatorMap.computeIfAbsent(taskId, myTaskId -> new ArrayList<>()).add(appenderator);
```

Reply:
Changed, thanks